### PR TITLE
Merge ToC Extension into core

### PR DESCRIPTION
**Pull request by marthacryan**
## References

As discussed in [this issue](https://github.com/jupyterlab/team-compass/issues/63) and in the weekly dev meetings, [the ToC extension](https://github.com/jupyterlab/jupyterlab-toc) seems to have enough support to be merged into jupyterlab core. 

## Code changes

To follow the conventions of other core packages, the ToC extension is split into a `toc` package and a `toc-extension` package. There are a few changes to the configurations and imports to follow conventions for packages / make the extension work in core. Otherwise it's importing the toc code "as-is" at the moment. 

## User-facing changes

This will cause the ToC extension to appear as an option in the left side panel along with the File Browser, Running Terminals and Kernels, etc.

### CODE REVIEW

File: packages/toc-extension/package.json [hunk 1]
```diff
@@ -0,0 +1,70 @@
+{
+  "name": "@jupyterlab/toc-extension",
+  "version": "4.0.0",
+  "private": false,
```
Comment (ajbozarth) [inline packages/toc-extension/package.json:4]: I think we can remove this line

File: packages/toc/.gitignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
+*.bundle.*
```
Comment (ajbozarth) [inline packages/toc/.gitignore:1]: remove this file

File: packages/toc/.prettierignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+node_modules
```
Comment (ajbozarth) [inline packages/toc/.prettierignore:1]: remove this file

File: packages/toc/.prettierrc [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (ajbozarth) [inline packages/toc/.prettierrc:1]: remove this file

File: packages/toc/LICENSE [hunk 5]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2017, Project Jupyter Contributors
```
Comment (ajbozarth) [inline packages/toc/LICENSE:1]: remove this file

File: packages/toc/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# jupyterlab-toc
```
Comment (ajbozarth) [inline packages/toc/README.md:1]: The relevant contents of this file should be moved to the docs and then remove this file

File: packages/toc/binder/postBuild [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
Comment (ajbozarth) [inline packages/toc/binder/postBuild:1]: remove this file

File: packages/toc/binder/requirements.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+jupyterlab>=1.0
```
Comment (ajbozarth) [inline packages/toc/binder/requirements.txt:1]: remove this file

File: packages/toc/notebooks/demo.ipynb [hunk 9]
```diff
@@ -0,0 +1,994 @@
+{
```
Comment (ajbozarth) [inline packages/toc/notebooks/demo.ipynb:1]: remove this file

File: packages/toc/package.json [hunk 10]
```diff
@@ -0,0 +1,79 @@
+{
+  "name": "@jupyterlab/toc",
+  "version": "4.0.0",
+  "private": false,
+  "description": "Table of Contents extension for JupyterLab",
+  "keywords": [
+    "jupyterlab"
+  ],
+  "homepage": "https://github.com/jupyterlab/jupyterlab-toc",
```
Comment (ajbozarth) [inline packages/toc/package.json:9]: same comments as the other `package.json` update urls and description, and remove the private line

File: packages/toc/tslint.json [hunk 11]
```diff
@@ -0,0 +1,109 @@
+{
```
Comment (ajbozarth) [inline packages/toc/tslint.json:1]: remove this file

File: packages/toc/yarn.lock [hunk 12]
```diff
@@ -0,0 +1,3041 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (ajbozarth) [inline packages/toc/yarn.lock:1]: remove this file

**Comment by telamonian** [comment]
Wow. Beatiful job preserving the history of the original project.

This should probably get its own test suite before it goes into core. Are there any tests from the existing extension that can be salvaged here?

**Comment by ajbozarth** [comment]
@telamonian all the toc code is in this PR, it didn't have tests in it's repo :(

**Comment by telamonian** [comment]
Oh. Then I guess the history thing is slightly less impressive. Though I do still think its nice to be able to pick up everyone's contributions.

But yeah, this needs tests

**Comment by blink1073** [comment]
@jupyterlab/committers let's avoid merging any other PRs until this goes in to avoid huge headaches.

**Comment by blink1073** [comment]
@telamonian I agree we need some tests but I don't want to hold up this particular PR for too long to avoid major merge conflicts.  Worst case we can disable this extension by default if we feel it isn't stable enough for 3.0.

**Comment by blink1073** [comment]
@marthacryan please let us know once you've finished addressing pressing concerns, and we can merge and iterate.  The docs failure is unrelated but there is a lint error (please run `jlpm lint` locally and commit the changes).

**Comment by marthacryan** [comment]
Fixed the remaining lint errors - I had to use unsafe react methods (such as `UNSAFE_componentWillReceiveProps`) to fix a couple of these errors, but this method is used in another core extension in Jupyterlab [here](https://github.com/jupyterlab/jupyterlab/blob/master/packages/extensionmanager/src/widget.tsx#L509), so I thought it'd be ok for now - let me know if I should fix that in this PR. 

I also changed the version of `toc` and `toc-extension` to `4.2.0-alpha.0`.

**Comment by jasongrout** [comment]
Just checking - did you change the collapsing functionality already to a signal or something, or is that something we'll need to do in another PR?

**Comment by marthacryan** [comment]
@jasongrout No I didn't include the collapsing signal in this PR (it's currently a setting that makes the collapsing disabled by default).

**Comment by blink1073** [comment]
I opened https://github.com/jupyterlab/jupyterlab/issues/8546 for tracking.

**Comment by blink1073** [comment]
The usage test is hanging again, merging.  Thanks @marthacryan!  This is an epic first contribution :smile:

**Comment by marthacryan** [comment]
@blink1073 Thanks you!

**Pull request merged**